File: src/r_app/titanic_model.R
# -----------------------------
# Titanic Logistic Regression (Base R)
# -----------------------------
cat("Loading Titanic training data...\n")
train <- read.csv("data/train.csv")
cat("Data loaded successfully!\n")

cat("Data preview:\n")
print(head(train, 5))

cat("\nBasic info:\n")
str(train)

cat("\nMissing values per column:\n")
print(colSums(is.na(train)))

# -----------------------------
# Clean and preprocess data
# -----------------------------
cat("\n🔧 Cleaning and preparing data...\n")

train$Age[is.na(train$Age)] <- median(train$Age, na.rm = TRUE)
train$Embarked[is.na(train$Embarked)] <- "S"
train$Sex <- ifelse(train$Sex == "male", 0, 1)

# Convert Embarked to dummy variables manually
train$Embarked_C <- ifelse(train$Embarked == "C", 1, 0)
train$Embarked_Q <- ifelse(train$Embarked == "Q", 1, 0)
train$Embarked_S <- ifelse(train$Embarked == "S", 1, 0)

# Select relevant columns
features <- c("Pclass", "Sex", "Age", "SibSp", "Parch", "Fare",
              "Embarked_C", "Embarked_Q", "Embarked_S")
train_data <- train[, c("Survived", features)]

# Split training and validation data (80/20)
set.seed(42)
n <- nrow(train_data)
train_idx <- sample(seq_len(n), size = 0.8 * n)
train_set <- train_data[train_idx, ]
val_set <- train_data[-train_idx, ]

# -----------------------------
# Train logistic regression
# -----------------------------
cat("\nTraining logistic regression model...\n")
model <- glm(Survived ~ ., data = train_set, family = binomial)
cat("Model trained successfully!\n")

# -----------------------------
# Evaluate performance
# -----------------------------
train_pred <- ifelse(predict(model, train_set, type = "response") > 0.5, 1, 0)
val_pred <- ifelse(predict(model, val_set, type = "response") > 0.5, 1, 0)

train_acc <- mean(train_pred == train_set$Survived)
val_acc <- mean(val_pred == val_set$Survived)

cat(sprintf("Training Accuracy: %.4f\n", train_acc))
cat(sprintf("Validation Accuracy: %.4f\n", val_acc))

# -----------------------------
# Predict on test data
# -----------------------------
cat("\nLoading test dataset...\n")
test <- read.csv("data/test.csv")
test$Age[is.na(test$Age)] <- median(train$Age, na.rm = TRUE)
test$Fare[is.na(test$Fare)] <- median(train$Fare, na.rm = TRUE)
test$Sex <- ifelse(test$Sex == "male", 0, 1)
test$Embarked[is.na(test$Embarked)] <- "S"
test$Embarked_C <- ifelse(test$Embarked == "C", 1, 0)
test$Embarked_Q <- ifelse(test$Embarked == "Q", 1, 0)
test$Embarked_S <- ifelse(test$Embarked == "S", 1, 0)

X_test <- test[, features]

cat("\nMaking predictions on test dataset...\n")
test$Survived_Pred <- ifelse(predict(model, X_test, type = "response") > 0.5, 1, 0)

cat("Sample predictions (PassengerId → Survived):\n")
print(head(test[, c("PassengerId", "Survived_Pred")], 10))

cat("\nPrediction complete! Titanic model finished successfully.\n")

File: src/python_app/titanic_model.py
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

print("Loading Titanic training data...")
df = pd.read_csv("data/train.csv")
print("Data loaded successfully!")

print("Data preview:")
print(df.head(), "\n")

print("Basic info:")
print(df.info(), "\n")

print("Missing values per column:")
print(df.isnull().sum(), "\n")

# -----------------------------
# Clean and preprocess the data
# -----------------------------
print("🔧 Cleaning and preparing data...")

# Fill missing Age and Embarked values
df["Age"].fillna(df["Age"].median(), inplace=True)
df["Embarked"].fillna(df["Embarked"].mode()[0], inplace=True)

# Encode categorical variables
df["Sex"] = df["Sex"].map({"male": 0, "female": 1})
df = pd.get_dummies(df, columns=["Embarked"], drop_first=True)

# Select features and target
features = ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked_Q", "Embarked_S"]
X = df[features]
y = df["Survived"]

# Split for validation
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# -----------------------------
# Build and train model
# -----------------------------

print("Training Logistic Regression model...")
model = LogisticRegression(max_iter=200)
model.fit(X_train, y_train)
print("Model trained successfully!\n")

# -----------------------------
# Evaluate on training and validation data
# -----------------------------
train_pred = model.predict(X_train)
val_pred = model.predict(X_val)

print(f"Training Accuracy: {accuracy_score(y_train, train_pred):.4f}")
print(f"Validation Accuracy: {accuracy_score(y_val, val_pred):.4f}\n")

# -----------------------------
# Load test data and predict
# -----------------------------
print("Loading test dataset...")
test_df = pd.read_csv("data/test.csv")

# Fill missing values in test set
test_df["Age"] = test_df["Age"].fillna(df["Age"].median())
test_df["Fare"] = test_df["Fare"].fillna(df["Fare"].median())

# Encode categorical variables in test data
test_df["Sex"] = test_df["Sex"].map({"male": 0, "female": 1})
test_df = pd.get_dummies(test_df, columns=["Embarked"], drop_first=True)

# Ensure test_df has all required columns (some test files miss Embarked_Q)
for col in features:
    if col not in test_df.columns:
        test_df[col] = 0

X_test = test_df[features]

print("Making predictions on test dataset...")
test_pred = model.predict(X_test)

# Display first few predictions
print("Sample predictions (PassengerId → Survived):")
for pid, pred in zip(test_df["PassengerId"][:10], test_pred[:10]):
    print(f"{pid}: {pred}")

print("\nPrediction complete! Titanic model finished successfully.")